server/ip-tools.ts

```diff
@@ -129,20 +133,20 @@ export const IPTools = new class {
 			num = (num - part) / 256;
 			ipParts.unshift(part.toString());
 		}
+		if (ipParts.length !== 4) return null;
 		return ipParts.join('.');
 	}
 
 	getCidrRange(cidr: string): AddressRange | null {
 		if (!cidr) return null;
 		const index = cidr.indexOf('/');
 		if (index <= 0) {
-			return {
-				minIP: IPTools.ipToNumber(cidr),
-				maxIP: IPTools.ipToNumber(cidr),
-			};
+			const ip = IPTools.ipToNumber(cidr);
+			if (ip < 0) return null;
+			return {minIP: ip, maxIP: ip};
 		}
 		const low = IPTools.ipToNumber(cidr.slice(0, index));
-		const bits = parseInt(cidr.slice(index + 1));
+		const bits = Utils.parseExactInt(cidr.slice(index + 1));
 		// fun fact: IPTools fails if bits <= 1 because JavaScript
 		// does << with signed int32s.
```

Comment:
```suggestion
		// fun fact: IPTools fails if bits <= 1 because JavaScript
		// does << with signed int32s.
		if (!bits || bits < 2 || bits > 32) return null;
```